05-type-classes/hw-05.hs

```diff
@@ -0,0 +1,254 @@
+Start out by creating a new data type VarExprT which is the same as ExprT but with an extra constructor for variables. Make VarExprT an instance of both Expr and HasVars. You should now be able to write things like
+ *Calc> add (lit 3) (var "x") :: VarExprT
+But we can’t stop there: we want to be able to interpret expres- sions containing variables, given a suitable mapping from variables to values. For storing mappings from variables to values, you should use the Data.Map module. Add
+import qualified Data.Map as M
+at the top of your file. The qualified import means that you must prefix M. whenever you refer to things from Data.Map. This is stan- dard practice, since Data.Map exports quite a few functions with names that overlap with names from the Prelude. Consult the Data.Map documentation to read about the operations that are sup- ported on Maps.
+Implement the following instances:
+instance HasVars (M.Map String Integer -> Maybe Integer) instance Expr (M.Map String Integer -> Maybe Integer)
+The first instance says that variables can be interpreted as func- tions from a mapping of variables to Integer values to (possibly) Integer values. It should work by looking up the variable in the mapping.
+The second instance says that these same functions can be inter- preted as expressions (by passing along the mapping to subexpres- sions and combining results appropriately).
+http://hackage.haskell.org/ packages/archive/containers/latest/ doc/html/Data- Map.html
+Note: to write these instances you will need to enable the FlexibleInstances language extension by putting
+cis 194: homework 5 6
+{-# LANGUAGE FlexibleInstances #-}
+as the first line in your file.
+Once you have created these instances, you should be able to test
+them as follows:
+withVars :: [(String, Integer)]
+-> (M.Map String Integer -> Maybe Integer) -> Maybe Integer
+withVars vs exp = exp $ M.fromList vs
+ *Calc> :t add (lit 3) (var "x")
+ add (lit 3) (var "x") :: (Expr a, HasVars a) => a
+ *Calc> withVars [("x", 6)] $ add (lit 3) (var "x")
+ Just 9
+ *Expr> withVars [("x", 6)] $ add (lit 3) (var "y")
+ Nothing
+ *Calc> withVars [("x", 6), ("y", 3)]
+Just 54
+-}
+--Exercise 6
+class HasVars a where
+    var :: String -> a
+
+data VarExprT = VarLit Integer
+    | VarAdd VarExprT VarExprT
+    | VarMul VarExprT VarExprT
+    | Var String
+    deriving (Show, Eq)
+
+instance Expr VarExprT where
+    lit = VarLit
+    mul = VarMul
+    add = VarAdd
+    
+instance HasVars VarExprT where
+    var = Var
+
+
+instance HasVars (M.Map String Integer -> Maybe Integer) where
+    var = M.lookup
+
+instance Expr (M.Map String Integer -> Maybe Integer) where
+    lit num = \_ -> Just num
+    mul leftFunc rightFunc = \map -> case (leftFunc map, rightFunc map) of
+        (Just firstVal, Just secondVal) -> Just (firstVal * secondVal)
```

Comment:
(Optional) I don't really remember if we had `Applicative` in previous homeworks, but this would be a perfect opportunity to use this to simplify this :smile:

Reply:
@jhrcek haha not yet I think, but I will check it out. Have seen it popping up elsewhere too.